### ISSUE TITLE
Swifter much slower than vanilla apply

### ISSUE THREAD

**Issue by apcamargo** [jmcarpenter2/swifter#51]
Hi @jmcarpenter2

I'm trying to use `swifter` to speed up a package I'm working on, but I'm getting some strange results. Using `swifter` is making the execution 3 to 10 times slower, depending on the function.

[test_swifter.pdf](https://github.com/jmcarpenter2/swifter/files/3077774/test_swifter.pdf)

**Comment by jmcarpenter2** [comment]
Hi @apcamargo,

Thanks for raising this issue and creating an easy to reproduce use case. This made understanding and debugging your issue MUCH easier!

So a couple things I noticed here. First, you are comparing the speed of swifter to vanilla pandas apply on a tiny data set. You are absolutely correct that swifter will lose every time on a data set of this size. The reason being is that the way swifter works is to first estimate the time of performing the apply on the first 1000 rows of the dataset 3 times and then do some extrapolation to estimate the time it would take to run on the full data set. At that point, it determines whether it is faster to do a dask apply or a pandas apply. In the case of your functions, it was faster to do a pandas apply because it took a matter of ~250 milliseconds to run the apply. Since your example contained <1000 rows, it actually ran a pandas apply 4 times total. Hence why the compute time for swifter is approximately 4-5 times the length of your vanilla pandas apply. 

This made me realize that I shouldn't necessarily use the first 1000 rows to estimate duration for tiny datasets. If there is less than 25000 rows, maybe we just want to estimate on 1/25 of the dataset. In this way we can nearly asymptote to pandas performance on tiny datasets as well.

Secondly, I noticed that on large datasets (700k rows) with your code, swifter was still using the pandas apply even though it should definitely be parallelizing with dask. It turns out that there was a missing pass of the keyword arguments to dask which caused the result_type='broadcast' to fail and the code to fallback to a pandas apply.

Both of these issues have been fixed in the subsequent PR, [#52](https://github.com/jmcarpenter2/swifter/pull/52), I will have merged into the package and released another version. Please `pip install swifter==0.286` or `pip install -U swifter` to get the latest version.

Thank you so much for helping me identify these former shortcomings of swifter so that I can continue to improve the package for all users.
Jason

**Comment by apcamargo** [comment]
You're welcome, @jmcarpenter2 ! I'm really excited to use swifter in a webapp I'm developing. I think it will be very useful to cut down processing time.

If you don't mind, I'd like to ask some questions about some design decisions you made with swifter, so I can understand it better.

- Why would pandas apply be faster than dask for small datasets? (this has nothing to do with swifter, but I'm ignorant about the inner working of dask).

- How did you chose the time threshold for dask? Wouldn't it be more logical for the threshold to be proportional to the size of the test set (as it no longer have a fixed size of 1000 rows)?

- Why did you chose to use 1/25 of the dataset as a test set? I would have thought that only a couple of rows would be enough.

I'm just starting to study pandas optimization, so I'm unsure if my questions are pertinent.

Again, thank you for your work and your detailed explanation of the issue!

**Comment by jmcarpenter2** [comment]
That's awesome @apcamargo, really cool to hear about use cases for swifter.

I will answer your questions in-line:

> Why would pandas apply be faster than dask for small datasets? (this has nothing to do with swifter, but I'm ignorant about the inner working of dask).

This is because there is a cost to scheduling the task graph and spinning up multiple processes to do the job. When the dataset is small enough, it is quicker to simply dive right into a pandas apply (which is basically just a for-loop) than to have all the overheard of partitioning the data and distributing it across workers and scheduling the jobs.

> How did you choose the time threshold for dask? Wouldn't it be more logical for the threshold to be proportional to the size of the test set (as it no longer have a fixed size of 1000 rows)?

This is a great question. I chose the 1 second time threshold based on two things: 1) A time estimation that each job that is spun up by the dask scheduler takes about 100ms to launch so 10 jobs should take just under a second to schedule and begin processing. 2) extensive experimentation when I was first developing the package. 

To be perfectly clear though, I am not convinced that 1 second is a universally ideal time threshold. It's more of a rough estimation to classify your problem into one of two broad buckets: small data/simple function and big data/complex function. Empirically, many problems are easily classified into one of those two buckets by the 1 second threshold.

However, I should add that it should not be relative to the size of the test set because the 1 second threshold is compared against the estimation of calling the pandas apply on the entire data set, not the just the test set.

> Why did you choose to use 1/25 of the dataset as a test set? I would have thought that only a couple of rows would be enough.

This is a pretty arbitrary choice, but the point here isn't to be exact but instead strike a balance between a small enough data set that we can estimate the full dataset function call quickly and a big enough data set that we can estimate the full dataset function call accurately through extrapolation.

To summarize the design choices in this package:
1. Start small and only run more expensive processes as they become necessary to speed up overall processing time
2. Use rough estimations to determine what is necessary and minimize overheard

Thanks for asking these questions, it's good to review the design choices (and consider making new ones as appropriate) occasionally.

-Jason

**Comment by apcamargo** [comment]
Thank you for the explanations, @jmcarpenter2 !
